### ISSUE TITLE
Should we use namespacing in the CLI

### ISSUE THREAD

**Issue by Tyriar** [microsoft/vscode#93241]
@Microsoft/vscode

https://github.com/microsoft/vscode/issues/92031 is requesting a terminal CLI to enable things like creating a new tab and splitting the current tab which would align with the capabilities of many terminals (including [Windows Terminal](https://github.com/microsoft/terminal/blob/master/doc/user-docs/UsingCommandlineArguments.md#using-the-wtexe-commandline)) and enable splitting a terminal tab by running a command in the terminal.

Does anyone have opinions for or against this? If we want to go with this I think we should also do the same for extensions by deprecating/aliasing the current CLI and moving them to either `code ext ...` or `code extension ...` as it keeps growing and could be better organized:

```
Extensions Management
  --extensions-dir <dir>                            Set the root path for
                                                    extensions.
  --list-extensions                                 List the installed
                                                    extensions.
  --show-versions                                   Show versions of installed
                                                    extensions, when using
                                                    --list-extension.
  --category                                        Filters installed
                                                    extensions by provided
                                                    category, when using
                                                    --list-extension.
  --install-extension <extension-id | path-to-vsix> Installs or updates the
                                                    extension. Use `--force`
                                                    argument to avoid
                                                    prompts.
  --uninstall-extension <extension-id>              Uninstalls an extension.
  --enable-proposed-api <extension-id>              Enables proposed API
                                                    features for extensions.
                                                    Can receive one or more
                                                    extension IDs to enable
                                                    individually.
```

**Comment by sbatten** [comment]
I really find CLIs that use this approach much easier to use. I would be in favor of having commands like `code ext install` over `code --install-extension` because they are easier to remember and type.

**Comment by bpasero** [comment]
I personally would prefer to not introduce any more capabilities via the CLI other than:
* opening something (this is what `code` is for)
* extension management: simply because this is probably a task that admins want to run headless and it does run headless
* troubleshooting (I think this is easy to justify)

**Comment by Tyriar** [comment]
> I personally would prefer to not introduce any more capabilities via the CLI other than

@bpasero why though?

All the proposed terminal ones (new tab, split pane, focus at/next/prev) would simply just route commands to the last active renderer so they could be built relatively easily. Extending upon this, the use case could also be enabled by a generic command CLI:

```
code --command workbench.action.terminal.split
code --command workbench.action.terminal.focusAtIndex 3
```

**Comment by roblourens** [comment]
> and enable splitting a terminal tab by running a command in the terminal

Why do people want to do that when it seems easier to go through the normal vscode command palette/UI?

**Comment by bpasero** [comment]
> @bpasero why though?

Couple of reasons, I personally find that when someone types `code --help` we should really try to not overwhelm the user with all the possibilities and show an absolute minimum set to get things going. 

On top of that, the CLI should drive the main capacity of the application, not some part within. You cannot even split the editor from the command line, so why the terminal?

**Comment by Tyriar** [comment]
@roblourens so you could for example have an alias to split the terminal (`alias cs=code terminal split-pane`) and so tools that do this for existing terminals can support vscode https://github.com/ranyitz/newsh/issues/11

> Couple of reasons, I personally find that when someone types code --help we should really try to not overwhelm the user with all the possibilities and show an absolute minimum set to get things going.

@bpasero That is exactly the problem this is trying to solve. We could change this:

```
Visual Studio Code - Insiders 1.44.0-insider

Usage: code-insiders.exe [options][paths...]

To read output from another program, append '-' (e.g. 'echo Hello World | code-insiders.exe -')

Options
  -d --diff <file> <file>           Compare two files with each other.
  -a --add <folder>                 Add folder(s) to the last active window.
  -g --goto <file:line[:character]> Open a file at the path on the specified
                                    line and character position.
  -n --new-window                   Force to open a new window.
  -r --reuse-window                 Force to open a file or folder in an
                                    already opened window.
  -w --wait                         Wait for the files to be closed before
                                    returning.
  --locale <locale>                 The locale to use (e.g. en-US or zh-TW).
  --user-data-dir <dir>             Specifies the directory that user data is
                                    kept in. Can be used to open multiple
                                    distinct instances of Code.
  -v --version                      Print version.
  -h --help                         Print usage.
  --telemetry                       Shows all telemetry events which VS code
                                    collects.
  --folder-uri <uri>                Opens a window with given folder uri(s)
  --file-uri <uri>                  Opens a window with given file uri(s)
  --sync <on> <off>                 Turn sync on or off

Extensions Management
  --extensions-dir <dir>                            Set the root path for
                                                    extensions.
  --list-extensions                                 List the installed
                                                    extensions.
  --show-versions                                   Show versions of installed
                                                    extensions, when using
                                                    --list-extension.
  --category                                        Filters installed
                                                    extensions by provided
                                                    category, when using
                                                    --list-extension.
  --install-extension <extension-id | path-to-vsix> Installs or updates the
                                                    extension. Use `--force`
                                                    argument to avoid
                                                    prompts.
  --uninstall-extension <extension-id>              Uninstalls an extension.
  --enable-proposed-api <extension-id>              Enables proposed API
                                                    features for extensions.
                                                    Can receive one or more
                                                    extension IDs to enable
                                                    individually.

Troubleshooting
  --verbose                          Print verbose output (implies --wait).
  --log <level>                      Log level to use. Default is 'info'.
                                     Allowed values are 'critical', 'error',
                                     'warn', 'info', 'debug', 'trace', 'off'.
  -s --status                        Print process usage and diagnostics
                                     information.
  --prof-startup                     Run CPU profiler during startup
  --disable-extensions               Disable all installed extensions.
  --disable-extension <extension-id> Disable an extension.
  --inspect-extensions <port>        Allow debugging and profiling of
                                     extensions. Check the developer tools for
                                     the connection URI.
  --inspect-brk-extensions <port>    Allow debugging and profiling of
                                     extensions with the extension host being
                                     paused after start. Check the developer
                                     tools for the connection URI.
  --disable-gpu                      Disable GPU hardware acceleration.
  --max-memory                       Max memory size for a window (in Mbytes).
```

Into this (more potential to move window management, troubleshooting into their own too):

```
Visual Studio Code - Insiders 1.44.0-insider

Usage: code-insiders.exe [options][paths...]

To read output from another program, append '-' (e.g. 'echo Hello World | code-insiders.exe -')

Options
  -d --diff <file> <file>           Compare two files with each other.
  -a --add <folder>                 Add folder(s) to the last active window.
  -g --goto <file:line[:character]> Open a file at the path on the specified
                                    line and character position.
  -n --new-window                   Force to open a new window.
  -r --reuse-window                 Force to open a file or folder in an
                                    already opened window.
  -w --wait                         Wait for the files to be closed before
                                    returning.
  --locale <locale>                 The locale to use (e.g. en-US or zh-TW).
  --user-data-dir <dir>             Specifies the directory that user data is
                                    kept in. Can be used to open multiple
                                    distinct instances of Code.
  -v --version                      Print version.
  -h --help                         Print usage.
  --telemetry                       Shows all telemetry events which VS code
                                    collects.
  --folder-uri <uri>                Opens a window with given folder uri(s)
  --file-uri <uri>                  Opens a window with given file uri(s)
  --sync <on> <off>                 Turn sync on or off

Commands:
  - extension
  - terminal

Troubleshooting
  --verbose                          Print verbose output (implies --wait).
  --log <level>                      Log level to use. Default is 'info'.
                                     Allowed values are 'critical', 'error',
                                     'warn', 'info', 'debug', 'trace', 'off'.
  -s --status                        Print process usage and diagnostics
                                     information.
  --prof-startup                     Run CPU profiler during startup
  --disable-extensions               Disable all installed extensions.
  --disable-extension <extension-id> Disable an extension.
  --inspect-extensions <port>        Allow debugging and profiling of
                                     extensions. Check the developer tools for
                                     the connection URI.
  --inspect-brk-extensions <port>    Allow debugging and profiling of
                                     extensions with the extension host being
                                     paused after start. Check the developer
                                     tools for the connection URI.
  --disable-gpu                      Disable GPU hardware acceleration.
  --max-memory                       Max memory size for a window (in Mbytes).
```

> On top of that, the CLI should drive the main capacity of the application, not some part within. You cannot even split the editor from the command line, so why the terminal?

If you're purely against driving the terminal and not namespaces in the CLI you should move your 👎 to https://github.com/microsoft/vscode/issues/92031

**Comment by bpasero** [comment]
@Tyriar ah sorry, makes sense. I think this could work but we need to somehow preserve backwards compatibility, no?

**Comment by Tyriar** [comment]
@bpasero yes backwards compat is essential, so `--install-extension ...` would be an alias for `extension install ...`. If it were just extension management then it wouldn't be worth it, but right now I don't feel like we can enhance the CLI any further as it clutters `code -h` too much.

**Comment by bpasero** [comment]
Back to the original proposal of doing this because people want to control the terminal via CLI, is that really a scenario people want? Are they in an integrated terminal, type `code split-terminal...` to split the terminal instead of just using the keyboard that would do this in a fraction of the time it needs to start a `Code.exe` process?

**Comment by Tyriar** [comment]
@bpasero there's a terminal-agnostic tool that does this which is where the requests initially came from eg. `newsh --split` (https://github.com/ranyitz/newsh/issues/11). Windows Terminal also had this request, they have a keybindings system to let you configure these as well.

- Docs https://github.com/microsoft/terminal/blob/master/doc/user-docs/UsingCommandlineArguments.md#using-the-wtexe-commandline
- PR https://github.com/microsoft/terminal/pull/4023
- Request (almost 100 :+1:) https://github.com/microsoft/terminal/issues/607

But no, I probably wouldn't use these and don't feel that passionate about championing the feature. And if we don't do that, then namespacing probably isn't a good idea until we come up with another category.

**Comment by bpasero** [comment]
I suggest to close this as "out of scope".

**Comment by bpasero** [comment]
I think that CLI option should not appear in the list of most used entries when running `--help`.

**Comment by aeschli** [comment]
Yes, we need some review and approval process. Feel free to assign me as reviewer.

@Tyriar We have `locate-extension` hat returns the location of an extension. Maybe call it --locate-shell-integration-dir

We only want the major aguments to be listed with `--help`. But I agree we need something like `--help --verbose` or `--help2` for the rest of the arguments.

**Comment by connor4312** [comment]
As the CLI does more things (server distro, "gateway" mode, version management, terminal work, other things in the future...) I do not think the current model of free-floating flags works very well. Note that when I say CLI here I refer to the interface exposed to users, which is eventually the Rust CLI.

Currently there is top-level help for extension management and flags that only apply to extension management. Adding similar sections for all other functionality causes help to become increasingly verbose and less useful to users, and I think a `--help --verbose` option does not address the root cause and is obscure to users. I suggest adopting a subcommand approach, where based editor options are shown at the top level (so `code --wait foo.txt` for instance still works) but the additional functionality I listed previously appears in a subcommand. If I'm a user who knows I want to do things with extensions, I can run `code ext --help` rather than trying to `code --help --verbose | grep -B 5 -A 5 -i extension`.

This is the approach I recently took in https://github.com/microsoft/vscode-cli/pull/432. That PR includes logic that handles things that became subcommands (`--(list|install|uninstall|)-extensions`, `--status`) in a backwards compatible way.
